datalake: various tweaks to shutdown

andrwng opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->
I have an upcoming PR that introduces some tests, and in debugging those tests, I that the datalake subsystems could be slow to shut down. Makes some tweaks in the area to address this, and adds some logging that made it easier to debug.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [x] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

andrwng committed:
datalake: debug logging on shutdown

It's helpful in debugging slow shutdowns to know when we begin and end
shutting down. Adds a couple debug logs.

andrwng committed:
pp/sr/api: tweaks to api shutdown

Adds debug logging, and tweaks lifecycle so that clients get shut down
before the service and sequencer.

Since the clients would previously be shutdown later, it could take some
time before the whole SR API shut down, causing flakiness in tests that
actively use SR during shutdown (e.g. datalake tests that are in the
middle of translating when shutdown is called).  Stopping down the
clients first avoids a slow node shutdown.

andrwng committed:
iceberg/rest_client: handle aborted sleep

We use expected<> throughout that area of code, and it's unexpected for
exceptions to be thrown. This handles exceptions that come from
sleep_abortable() and has it return an appropriate result. We could
previously run into an error log during shutdown:

```
ERROR 2025-02-26 23:16:40,520 [shard 0:main] datalake - coordinator.cc:124 - Coordinator hit exception while running in term 1: seastar::sleep_aborted (Sleep is aborted)
```

andrwng commented:
/ci-repeat

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5f2d69d5-99bc-4837-a1f8-8fb89151ee01-->
<!--TEST RESULTS-->
<!--331b0f67-b545-40be-be85-4364954f387b-->
<!--TEST RESULTS-->
<!--4720704e-e1e7-4019-8b32-321643c50811-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62380">build#62380</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/62380#019545a3-a358-4a1b-9a32-86528639982d | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62391">build#62391</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/62391#019546ab-3957-4406-84b3-31aef9a0e967 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/62391#019546ab-3959-46c2-82fb-1599117f0af5 | FLAKY | 1/2 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsets.test_consumer_group_offsets | ducktape | https://buildkite.com/redpanda/redpanda/builds/62391#019546a8-a325-499e-9fda-1a0edd876175 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62423">build#62423</a></summary>


| test_id                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.maintenance_test.MaintenanceTest.test_maintenance_sticky.use_rpk=True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62423#019548bc-8116-4b6d-ac47-0b1312b46a6f | FLAKY         | 1/2      |
</details>
<!--/body-->

andrwng committed:
iceberg: stop catalog client on shutdown

Previously the coordinator manager shutdown could take a while, notably
when a coordinator was in the middle of calling into the catalog. This
would make some tests with Iceberg topics timeout when shutting down.

This updates the catalog implementations to define a stop() method
(already defined for the REST catalog, but never called previously), and
makes the coordinator manager use it upon shutting down.

nvartolomei commented on src/v/datalake/datalake_manager.cc:
```diff
@@ -263,15 +263,17 @@ ss::future<> datalake_manager::start() {
       });
-    co_await std::move(f);
     _schema_cache->stop();
+    co_await std::move(f);
```
I don't understand what changes by swapping these 2 lines. The `gate.close()` task is already running since scheduled on line 267. Whether or not you waited on the future to resolve does not have an influence of how the underlying state looks except that now stop() method is called slightly faster and possibly you are just masking a race condition?

andrwng replied:
Ah that's fair. I think I went overboard here thinking it's never safe to access state outside of a gate. But in a stop() method it's probably fine. Will remove this aspect of this PR

andrwng replied:
Removed, and squashed the remaining logging-only commits together

nvartolomei reviewed (approved):
Appreciate detailed commit messages! Makes reviewing a breeze.

oleiman commented on src/v/iceberg/rest_client/catalog_client.cc:
```diff
@@ -224,7 +225,14 @@ ss::future<expected<iobuf>> catalog_client::perform_request(
-        co_await ss::sleep_abortable(permit.delay, rtc.root_abort_source());
+        auto sleep_fut = co_await ss::coroutine::as_future(
+          ss::sleep_abortable(permit.delay, rtc.root_abort_source()));
+        if (sleep_fut.failed()) {
```
don't you need to call `get` on the non-failure path?

andrwng replied:
Mm I don't think so? There's nothing in the result in the non-failure path because it's a `future<void>`. At least, it hasn't been complaining about an ignored future, if that's the concern, and this pattern does appear in at least a couple places in the codebase. Have you seen it be problematic?

oleiman replied:
ah yeah, i got it turned around in my head. confirmed that only exceptional futures need to be explicitly ignored.

oleiman commented on src/v/iceberg/filesystem_catalog.h:
```diff
@@ -95,6 +95,8 @@ class filesystem_catalog : public catalog {
     ss::future<checked<std::nullopt_t, errc>>
     commit_txn(const table_identifier& table_ident, transaction) final;
 
+    ss::future<> stop() final { return ss::make_ready_future(); }
```
micro-nit (please don't bother changing)

```suggestion
    ss::future<> stop() final { return ss::now(); }
```

```bash
$ git grep "make_ready_future()" | wc -l
19
$ git grep "ss::now" | wc -l
349
```

oleiman reviewed (approved):
lgtm
